builtin/k8s/task.go

```diff
@@ -402,9 +421,160 @@ func (p *TaskLauncher) StartTask(
+		ns = p.config.Namespace
+	}
+
+	// List pods with this job label
+	podsClient := clientSet.CoreV1().Pods(ns)
+	pods, err := podsClient.List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("job-name=%s", ti.Id),
+	})
+	// It's not clear from the documentation if an error is returned from the
+	// List API call if no jobs are found, so we guard here just in case
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if pods == nil {
+		log.Info("no pods found for job, returning", "job_id", ti.Id)
+		return nil, nil
+	}
+
+	// Assume first one exists for now? Our task launcher for k8s only launches
+	// one pod right now.
+	pod := pods.Items[0]
+
+	// How long to wait for the pod in question to start before attempting to stream
+	// its logs.
+	podStartUpTimeout := time.Duration(30 * time.Second)
+	if p.config.WatchTaskStartupTimeoutSeconds != 0 {
+		podStartUpTimeout = time.Duration(p.config.WatchTaskStartupTimeoutSeconds) * time.Second
+	}
+
+	log.Info("waiting for pod to start", "name", pod.Name)
+
+	// Ensure the pod exists before attempting to stream its logs
+	err = wait.PollImmediate(time.Second, podStartUpTimeout, func() (bool, error) {
+		p, err := clientSet.CoreV1().Pods(ns).Get(ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		switch p.Status.Phase {
+		case v1.PodRunning, v1.PodFailed, v1.PodSucceeded:
+			return true, nil
+		case v1.PodPending, v1.PodUnknown:
+			return false, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		log.Error("pod failed to start before timeout", "timeout", podStartUpTimeout, "err", err)
+		return nil, err
+	}
+
+	log.Info("attempting to stream pod logs")
+
+	// Accumulate our result on this
+	var result component.TaskResult
+
+	req := clientSet.CoreV1().Pods(pod.Namespace).GetLogs(pod.Name, &corev1.PodLogOptions{
+		Follow: true,
+	})
+	podLogs, err := req.Stream(ctx)
+	if err != nil {
+		log.Error("failed to read pod log stream", "err", err)
+		return nil, err
+	}
+	defer podLogs.Close()
+
+	log.Info("reading pod logs", "name", pod.Name)
+
+	// Start a func to watch the pod phase. If complete, we no longer need to
+	// attempt to stream the pod logs.
+	logsDoneCh := make(chan bool)
+	go func() {
+		defer close(logsDoneCh)
+		for {
+			p, err := clientSet.CoreV1().Pods(ns).Get(ctx, pod.Name, metav1.GetOptions{})
+			if err != nil {
+				log.Warn("error getting pod status", "err", err)
+				ui.Output("Error getting pod status: %s", err, terminal.WithErrorStyle())
+
+				logsDoneCh <- true
+				return
+			}
+
+			switch p.Status.Phase {
+			case v1.PodRunning:
+				// Pod is still running, so wait
+				logsDoneCh <- false
+			case v1.PodFailed:
+				// Pod has finished
+				result.ExitCode = 1
+				logsDoneCh <- true
+			case v1.PodSucceeded:
+				// Pod has finished
+				result.ExitCode = 0
+				logsDoneCh <- true
+				return
+			case v1.PodPending, v1.PodUnknown:
+				// Unknown state, still wait
+				logsDoneCh <- false
+			}
```

Comment:
We can put a sleep here and exit on context cancel so that we're not hammering the k8s API looking at the pod phase.